internal/transport/grpchttp2/framer.go

```diff
@@ -0,0 +1,246 @@
+	FrameTypePriority     FrameType = 0x2
+	FrameTypeRSTStream    FrameType = 0x3
+	FrameTypeSettings     FrameType = 0x4
+	FrameTypePushPromise  FrameType = 0x5
+	FrameTypePing         FrameType = 0x6
+	FrameTypeGoAway       FrameType = 0x7
+	FrameTypeWindowUpdate FrameType = 0x8
+	FrameTypeContinuation FrameType = 0x9
+)
+
+// Flags represents one or more flags set on an HTTP/2 Frame.
+type Flags uint8
+
+const (
+	FlagDataEndStream          Flags = 0x1
+	FlagDataPadded             Flags = 0x8
+	FlagHeadersEndStream       Flags = 0x1
+	FlagHeadersEndHeaders      Flags = 0x4
+	FlagHeadersPadded          Flags = 0x8
+	FlagHeadersPriority        Flags = 0x20
+	FlagSettingsAck            Flags = 0x1
+	FlagPingAck                Flags = 0x1
+	FlagContinuationEndHeaders Flags = 0x4
+)
+
+// Setting represents the id and value pair of an HTTP/2 setting.
+// See [Setting Format].
+//
+// [Setting Format]: https://httpwg.org/specs/rfc7540.html#SettingFormat
+type Setting struct {
+	ID    SettingID
+	Value uint32
+}
+
+// SettingID represents the id of an HTTP/2 setting.
+// See [Setting Values].
+//
+// [Setting Values]: https://httpwg.org/specs/rfc7540.html#SettingValues
+type SettingID uint16
+
+const (
+	SettingsHeaderTableSize      SettingID = 0x1
+	SettingsEnablePush           SettingID = 0x2
+	SettingsMaxConcurrentStreams SettingID = 0x3
+	SettingsInitialWindowSize    SettingID = 0x4
+	SettingsMaxFrameSize         SettingID = 0x5
+	SettingsMaxHeaderListSize    SettingID = 0x6
+)
+
+// FrameHeader is the 9 byte header of any HTTP/2 Frame.
+// See [Frame Header].
+//
+// [Frame Header]: https://httpwg.org/specs/rfc7540.html#FrameHeader
+type FrameHeader struct {
+	// Size is the size of the frame's payload without the 9 header bytes.
+	// As per the HTTP/2 spec, size can be up to 3 bytes, but only frames
+	// up to 16KB can be processed without agreement.
+	Size uint32
+	// Type is a byte that represents the Frame Type.
+	Type FrameType
+	// Flags is a byte representing the flags set on this Frame.
+	Flags Flags
+	// StreamID is the ID for the stream which this frame is for. If the
+	// frame is connection specific instead of stream specific, the
+	// streamID is 0.
+	StreamID uint32
+}
+
+// Frame represents an HTTP/2 Frame.
+type Frame interface {
+	Header() *FrameHeader
+}
+
+type DataFrame struct {
+	hdr  *FrameHeader
+	free func()
+	Data []byte
+}
+
+func (f *DataFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *DataFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+type HeadersFrame struct {
+	hdr      *FrameHeader
+	free     func()
+	HdrBlock []byte
+}
+
+func (f *HeadersFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+func (f *HeadersFrame) Free() {
+	if f.free != nil {
+		f.free()
+	}
+}
+
+type RSTStreamFrame struct {
+	hdr  *FrameHeader
+	Code ErrCode
+}
+
+func (f *RSTStreamFrame) Header() *FrameHeader {
+	return f.hdr
+}
+
+type SettingsFrame struct {
+	hdr      *FrameHeader
+	free     func()
+	settings []Setting
```

Comment:
Should this field be exported like on the other frames?

Reply:
My rationale was to have it unexported and then provide some utility methods like the ones on the `net/x` implementation of the `SettingsFrame`. See [this](https://cs.opensource.google/go/x/net/+/master:http2/frame.go;l=761-769?q=type%20SettingsFrame&ss=go%2Fx%2Fnet) for reference. I didn't implement them since they are not used yet. I see three options:
* Implement the missing methods now
* Leave as is and implement the methods as needed
* Make the field exported and let the caller worry about getting the settings they need

What do you think? Do you see any other approach?

Reply:
> Leave as is and implement the methods as needed

Add a TODO saying implement the methods as needed. Thanks.